src/com/aaronzalewski/HtmlDocumentImpl.java

```diff
@@ -0,0 +1,85 @@
+package com.aaronzalewski;
+
+import com.google.common.collect.ImmutableList;
+import org.jsoup.Connection;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class HtmlDocumentImpl implements HtmlDocument {
+
+    private static final String EMAIL_REGEX = "[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+";
+
+    private String domain;
+    private Document document;
+
+    public static HtmlDocument createHtmlDocument(String url) {
+        try {
+            return new HtmlDocumentImpl(url);
+        } catch (IOException | IllegalArgumentException e) {
+            return new EmptyHtmlDocumentImpl();
+        }
+    }
+
+    private HtmlDocumentImpl(String url) throws IOException {
+        Connection connection = Jsoup.connect(url);
+        this.document = connection.get();
+        this.domain = getUrlDomain(url);
+    }
+
+    @Override
+    public List<String> getEmailAddresses() {
+        ImmutableList.Builder<String> emails = ImmutableList.builder();
```

Comment:
I decided to use an immutable list because the list represents the emails found on the page, which shouldn't change within the context of the application.  Now that I think about it though, an immutable set could be used here as well as in getUrlsInDomain(), since I don't want duplicates for either.